mount/mount_linux.go

```diff
@@ -289,55 +288,59 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
+	existingFormat, err := mounter.GetDiskFormat(source)
+	if err != nil {
+		return err
+	}
+
+	if existingFormat == "" {
+		// Do not attempt to format the disk if mounting as readonly, return an error to reflect this.
+		if readOnly {
+			return fmt.Errorf("Cannot mount unformatted disk %s as we are manipulating it in read-only mode", source)
 		}
-		if existingFormat == "" {
-			if readOnly {
-				// Don't attempt to format if mounting as readonly, return an error to reflect this.
-				return errors.New("failed to mount unformatted volume as read only")
-			}
 
-			// Disk is unformatted so format it.
-			args := []string{source}
-			// Use 'ext4' as the default
-			if len(fstype) == 0 {
-				fstype = "ext4"
-			}
+		// Disk is unformatted so format it.
+		args := []string{source}
+		// Use 'ext4' as the default
+		if len(fstype) == 0 {
+			fstype = "ext4"
+		}
 
-			if fstype == "ext4" || fstype == "ext3" {
-				args = []string{
-					"-F",  // Force flag
-					"-m0", // Zero blocks reserved for super-user
-					source,
-				}
+		if fstype == "ext4" || fstype == "ext3" {
+			args = []string{
+				"-F",  // Force flag
+				"-m0", // Zero blocks reserved for super-user
+				source,
 			}
-			klog.Infof("Disk %q appears to be unformatted, attempting to format as type: %q with options: %v", source, fstype, args)
-			_, err := mounter.Exec.Command("mkfs."+fstype, args...).CombinedOutput()
-			if err == nil {
-				// the disk has been formatted successfully try to mount it again.
-				klog.Infof("Disk successfully formatted (mkfs): %s - %s %s", fstype, source, target)
-				return mounter.Interface.Mount(source, target, fstype, options)
-			}
-			klog.Errorf("format of disk %q failed: type:(%q) target:(%q) options:(%q)error:(%v)", source, fstype, target, options, err)
-			return err
 		}
-		// Disk is already formatted and failed to mount
-		if len(fstype) == 0 || fstype == existingFormat {
-			// This is mount error
-			return mountErr
+
+		klog.Infof("Disk %q appears to be unformatted, attempting to format as type: %q with options: %v", source, fstype, args)
+		_, err := mounter.Exec.Command("mkfs."+fstype, args...).CombinedOutput()
+		if err == nil {
+			// the disk has been formatted successfully try to mount it again.
+			klog.Infof("Disk successfully formatted (mkfs): %s - %s %s", fstype, source, target)
+			return mounter.Interface.Mount(source, target, fstype, options)
 		}
-		// Block device is formatted with unexpected filesystem, let the user know
-		return fmt.Errorf("failed to mount the volume as %q, it already contains %s. Mount error: %v", fstype, existingFormat, mountErr)
+
+		klog.Errorf("format of disk %q failed: type:(%q) target:(%q) options:(%q)error:(%v)", source, fstype, target, options, err)
+		return err
+	} else {
+		klog.V(4).Infof("%s is already formatted (%s)", source, existingFormat)
+	}
+
+	// Verify that the disk is formatted with filesystem type we are expecting
+	if fstype != existingFormat {
```

Comment:
I think moving it to above "else" will be better.
```
if existingFormat == "" {
    ...
} else if existingFormat != fstype {
    ...
}
```
And also, setting fstype's default value before this if-else, so that it always has a defalut value.

Reply:
Indeed, I amended!